File: src/main/cs/Problem0024.cs
using System;
using System.Diagnostics;
using Euler.test.cs;

namespace Euler.main.cs
{
    public class Problem0024
    {
        private int _counter;

        public Problem0024()
        {
            Stopwatch stopWatch = new Stopwatch();
            stopWatch.Start();

            // Solution:
            int n = 1000000;
            string[] permutations = new string[n];
            string first = "0123456789";

            Permute(permutations, "", first);


            stopWatch.Stop();
            string elapsedTime = stopWatch.Elapsed.ToString();
            Console.WriteLine("Result:\t" + permutations[n - 1] + "\tTime:\t" +
                              (double.Parse(elapsedTime.Substring(elapsedTime.LastIndexOf(":") + 1, 2)) >= 1
                                  ? double.Parse(elapsedTime.Substring(elapsedTime.LastIndexOf(":") + 1)) + " s"
                                  : double.Parse(elapsedTime.Substring(elapsedTime.IndexOf(".") + 1)) / 10_000 +
                                    " ms"));
            if (Test.DoBenchmark)
                Benchmark.AddTime(24, double.Parse(elapsedTime.Substring(elapsedTime.IndexOf(".") + 1)) / 10_000);
        }

        private void Permute(string[] permutations, string prefix, string s)
        {
            if (_counter >= 1000000) return;

            int n = s.Length;
            if (n == 0) permutations[_counter++] = prefix;
            else
                for (int i = 0; i < n; i++)
                    Permute(permutations, prefix + s[i], s.Substring(0, i) + s.Substring(i + 1, n - i - 1));
        }
    }
}

File: src/main/java/de/swerik/euler/_00/Problem0024.java
package de.swerik.euler._00;

public class Problem0024 {

    private int counter = 0;

    public Problem0024() {
        long startTime = System.nanoTime();
        String result = "";

        // Solution:
        int n = 1000000;
        String[] permutations = new String[n];
        String first = "0123456789";

        this.permute(permutations, "", first);

        long timeToResolve = System.nanoTime() - startTime;
        System.out.println("Result:\t" + permutations[n - 1] + "\tTime:\t" + (((double) timeToResolve / 1_000_000) > 1000 ?
                (((double) timeToResolve / 1_000_000_000) + "s") :
                (((double) timeToResolve / 1_000_000) + "ms")));
    }

    private void permute(String[] permutations, String prefix, String s) {
        if (this.counter >= 1000000) {
            return;
        }
        int n = s.length();
        if (n == 0) {
            permutations[this.counter++] = prefix;
        } else {
            for (int i = 0; i < n; i++) {
                this.permute(permutations, prefix + s.charAt(i), s.substring(0, i) + s.substring(i + 1, n));
            }
        }
    }
}
